src/background/injector-controller.ts

```diff
@@ -52,9 +57,22 @@ export class InjectorController {
                         tabId: tabId,
                     }).result;
                 })
-                .catch(this.logger.error);
+                .catch(this.handleInjectionError);
         }
 
         this.oldInspectType = inspectStoreState.inspectMode;
     };
+
+    private handleInjectionError = async (err: any): Promise<void> => {
+        this.logger.error(err);
+        const attempts = (this.visualizationStore.getState().injectionAttempts ?? 0) + 1;
+        const payload = {
+            failedAttempts: attempts,
+            injectionFailed: attempts > 3,
+        } as InjectionFailedPayload;
```

Comment:
It would be preferable to use `const payload: InjectionFailedPayload = { /* contents */ };` here rather than `const payload = { /* contents */ } as InjectionFailedPayload;`. The former tells typescript "this variable may only be assigned objects which are of type `InjectionFailedPayload`", where the latter tells typescript "cast the object I've created here to an `InjectionFailedPayload` type, and infer the type of the variable to be the same type".

In practice, the latter gives fewer type protections because the `as` cast is telling typescript "yes, really treat this object as the type I'm telling you, *even if it's missing required properties*". In the former pattern, forgetting a required property (eg, accidentally omitting `injectionFailed` here) is an error, but in the latter pattern with `as`, it's not an error. That's why you often see the `as` pattern in unit tests, where we're often intentionally giving underdefined inputs (eg, `deps` that are intentionally missing unnecessary bits). But it's usually a bug to use `as` outside of unit tests - if a non-unit test really does need to omit part of a type, it's usually better to update the type to indicate that that part of the type might be null/undefined.

Reply:
Sounds good, thanks for the detailed explanation, that makes sense!